File: ios/Classes/Scanner Core/optimizer/CardDetailsScanOptimizer.swift
//
//  CardDetailsScanOptimizer.swift
//  card_scanner
//
//  Created by Abhinav Kumar on 12/02/21.
//

import Foundation

class CardDetailsScanOptimizer {
    private var scannerOptions: CardScannerOptions
    
    private var cardNumberFrequencyTable: [String: Int] = [:]
    private var expiryDateFrequencyTable: [String: Int] = [:]
    private var cardHolderNameFrequencyTable: [String: Int] = [:]
    private var optimalCardNumber: String? = nil
    private var optimalExpiryDate: String? = nil
    private var optimalCardHolderName: String? = nil
    private var numberOfCardDetailsProcessed: Int = 0
    
    init(scannerOptions: CardScannerOptions) {
        self.scannerOptions = scannerOptions
    }
    
    func processCardDetails(cardDetails: CardDetails) {
        guard cardDetails.cardNumber.isNotEmpty else { return }
        
        numberOfCardDetailsProcessed += 1
        
        // drop first few scan results which have more chances of errors
        guard numberOfCardDetailsProcessed > scannerOptions.initialScansToDrop else { return }
        
        handle(cardNumber: cardDetails.cardNumber)
        handle(expiryDate: cardDetails.expiryDate)
        handle(cardHolderName: cardDetails.cardHolderName)
        
        updateOptimalData()
        printStatus()
    }
    
    func isReadyToFinishScan() -> Bool {
        return numberOfCardDetailsProcessed > scannerOptions.validCardsToScanBeforeFinishingScan
    }
    
    private func updateOptimalData() {
        optimalCardNumber = cardNumberFrequencyTable.mostFrequentData
        optimalExpiryDate = expiryDateFrequencyTable.mostFrequentData
        optimalCardHolderName = cardHolderNameFrequencyTable.mostFrequentData
    }
    
    private func handle(cardNumber: String?) {
        guard let cardNumber = cardNumber, cardNumber.isNotEmpty else { return }
        
        let sanitizedCardNumber = cardNumber.sanitized.cardNumberSized
        
        cardNumberFrequencyTable[sanitizedCardNumber] = (cardNumberFrequencyTable[sanitizedCardNumber] ?? 0) + 1
    }
    
    private func handle(expiryDate: String?) {
        guard let expiryDate = expiryDate, expiryDate.isNotEmpty else { return }
    
        expiryDateFrequencyTable[expiryDate] = (expiryDateFrequencyTable[expiryDate] ?? 0) + 1
    }
    
    private func handle(cardHolderName: String?) {
        guard let cardHolderName = cardHolderName, cardHolderName.isNotEmpty else { return }
        
        cardHolderNameFrequencyTable[cardHolderName] = (cardHolderNameFrequencyTable[cardHolderName] ?? 0) + 1
    }
    
    func printStatus() {
        debugLog("Card Number : \(optimalCardNumber ?? "Not Yet Scanner")", scannerOptions: scannerOptions)
        debugLog("Card Number Freq : \(cardNumberFrequencyTable[optimalCardNumber ?? ""] ?? 0)", scannerOptions: scannerOptions)
        debugLog("Expiry Date =  \(optimalExpiryDate ?? "Not Yet Scanner")", scannerOptions: scannerOptions)
        debugLog("Expiry Date Freq : \(expiryDateFrequencyTable[optimalExpiryDate ?? ""] ?? 0)", scannerOptions: scannerOptions)
        debugLog("Card Holder Name :  \(optimalCardHolderName ?? "Not Yet Scanner")", scannerOptions: scannerOptions)
        debugLog("Card Holder Name Freq :  \(cardHolderNameFrequencyTable[optimalCardHolderName ?? ""] ?? 0)", scannerOptions: scannerOptions)
        debugLog("Scanner Options = \(scannerOptions)", scannerOptions: scannerOptions)
    }
    
    func getOptimalCardDetails() -> CardDetails? {
        guard let optimalCardNumber = optimalCardNumber else {
            return nil
        }
        
        printStatus()
        
        return CardDetails(
            cardNumber: optimalCardNumber,
            cardHolderName: optimalCardHolderName ?? "",
            expiryDate: optimalExpiryDate ?? ""
        )
    }
}

File: android/src/main/kotlin/com/basys/card_scanner/scanner_core/optimizer/CardDetailsScanOptimizer.kt
package com.basys.card_scanner.scanner_core.optimizer

import com.basys.card_scanner.logger.debugLog
import com.basys.card_scanner.scanner_core.models.CardDetails
import com.basys.card_scanner.scanner_core.models.CardScannerOptions

class CardDetailsScanOptimizer(private val _scannerOptions: CardScannerOptions) {
  private val _expiryDateFrequency: MutableMap<String, Int> = mutableMapOf()
  private val _cardNumberFrequency: MutableMap<String, Int> = mutableMapOf()
  private val _cardHolderNameFrequency: MutableMap<String, Int> = mutableMapOf()
  private var _optimalCardNumber: String? = null
  private var _optimalExpiryDate: String? = null
  private var _optimalCardHolderName: String? = null
  private var numberOfCardDetailsProcessed: Int = 0

  fun processCardDetails(cardDetails: CardDetails) {
    if (cardDetails.cardNumber.isEmpty()) return
    val cardNumber = cardDetails.cardNumber
    val expiryDate = cardDetails.expiryDate
    val cardHolder = cardDetails.cardHolderName
    numberOfCardDetailsProcessed++

    ///drop first few scan results which have more chances of errors
    if (numberOfCardDetailsProcessed <= _scannerOptions.initialScansToDrop) return
    handleCardNumber(cardNumber)
    handleExpiryDateNumber(expiryDate)
    handleCardHolderName(cardHolder)
    updateOptimalData()
  }

  fun isReadyToFinishScan(): Boolean {
    return numberOfCardDetailsProcessed > _scannerOptions.validCardsToScanBeforeFinishingScan
  }

  private fun updateOptimalData() {
    _optimalCardNumber = getMostFrequentData(_cardNumberFrequency)
    _optimalExpiryDate = getMostFrequentData(_expiryDateFrequency)
    _optimalCardHolderName = getMostFrequentData(_cardHolderNameFrequency)
  }

  private fun handleCardNumber(cardNumber: String) {
    if (cardNumber.isEmpty()) return
    _cardNumberFrequency[cardNumber] = (_cardNumberFrequency[cardNumber] ?: 0) + 1
  }

  private fun handleExpiryDateNumber(expiryDate: String) {
    if (expiryDate.isEmpty()) return
    _expiryDateFrequency[expiryDate] = (_expiryDateFrequency[expiryDate] ?: 0) + 1
  }

  private fun handleCardHolderName(cardHolderName: String) {
    if (cardHolderName.isEmpty()) return
    _cardHolderNameFrequency[cardHolderName] = (_cardHolderNameFrequency[cardHolderName] ?: 0) + 1
  }

  private fun getMostFrequentData(frequencyMap: MutableMap<String, Int>): String? {
    var mostFrequentEntry: Map.Entry<String, Int>? = null
    for (entry in frequencyMap.entries) {
      if (mostFrequentEntry == null || entry.value >= mostFrequentEntry.value) {
        mostFrequentEntry = entry
      }
    }
    return mostFrequentEntry?.key
  }

  private fun printStatus() {
    debugLog(" card number : " + _cardNumberFrequency[_optimalCardNumber
            ?: ""] + " , expiry = " + _expiryDateFrequency[_optimalExpiryDate
            ?: ""] + " , holder name = " + _cardHolderNameFrequency[_optimalCardHolderName], _scannerOptions)
  }

  fun getOptimalCardDetails(): CardDetails? {
    if (_optimalCardNumber == null) return null
    printStatus()
    return CardDetails(cardNumber = _optimalCardNumber!!, cardHolderName = _optimalCardHolderName
            ?: "", expiryDate = _optimalExpiryDate ?: "")
  }
}